src/v/cloud_storage/materialized_resources.cc

```diff
@@ -115,84 +105,159 @@ ss::future<> materialized_resources::start() {
+    auto units = s._units.count();
+    if (units > estimate) {
+        vlog(
+          cst_log.debug,
+          "Returning extra units. Current: {}, extimate: {}",
+          units,
+          estimate);
+        s._units.return_units(units - estimate);
+    } else {
+        vlog(
+          cst_log.debug,
+          "Adopting extra units. Current: {}, extimate: {}",
+          units,
+          estimate);
+        auto tr = _mem_units.take(estimate - units);
+        s._units.adopt(std::move(tr.units));
+        if (tr.checkpoint_hint) {
+            // We hit the memory limit and the semaphore counter
+            // went below zero. We need to mark this segment as a
+            // candidate for eviction.
+            s.overcommit = true;
+            vlog(
+              cst_log.info,
+              "{} Overcommit detected, segment overshoot by {}",
+              s.segment->get_ntp(),
+              estimate - units);
+        }
+    }
     _materialized.push_back(s);
 }
 
 ss::future<segment_reader_units>
 materialized_resources::get_segment_reader_units() {
-    if (_segment_reader_units.available_units() <= 0) {
+    // Estimate segment reader memory requirements
+    auto size_bytes = projected_remote_segment_reader_memory_usage();
+    if (_mem_units.available_units() <= size_bytes) {
         // Update metrics counter if we are trying to acquire units while
         // saturated
         _segment_readers_delayed += 1;
 
         trim_segment_readers(max_segment_readers() / 2);
     }
 
-    auto semaphore_units = co_await _segment_reader_units.get_units(1);
+    auto semaphore_units = co_await _mem_units.get_units(size_bytes);
     co_return segment_reader_units{std::move(semaphore_units)};
 }
 
 ss::future<ssx::semaphore_units>
-materialized_resources::get_partition_reader_units(size_t n) {
-    if (_partition_reader_units.available_units() <= 0) {
+materialized_resources::get_partition_reader_units() {
+    auto sz = projected_remote_partition_reader_memory_usage();
+    if (_mem_units.available_units() <= sz) {
         // Update metrics counter if we are trying to acquire units while
         // saturated
         _partition_readers_delayed += 1;
     }
-    return _partition_reader_units.get_units(n);
+    return _mem_units.get_units(sz);
 }
 
 ss::future<segment_units> materialized_resources::get_segment_units() {
-    if (_segment_units.available_units() <= 0) {
+    auto sz = projected_remote_segment_memory_usage();
+    if (_mem_units.available_units() <= sz) {
         // Update metrics counter if we are trying to acquire units while
         // saturated
         _segments_delayed += 1;
 
         trim_segments(max_segments() / 2);
     }
-    auto semaphore_units = co_await _segment_units.get_units(1);
+    auto semaphore_units = co_await _mem_units.get_units(sz);
     co_return segment_units{std::move(semaphore_units)};
 }
 
 void materialized_resources::trim_segment_readers(size_t target_free) {
```

Comment:
Does it make sense to have separate trims after this change? Maybe they should be unified.